web/modules/weather_data/src/Service/WeatherAlertTrait.php

```diff
@@ -190,6 +198,259 @@ public function getAlerts($grid, $point, $self = false)
 
         $this->cache->set($CACHE_KEY, $alerts, time() + 30);
 
+        $this->stashedAlerts = $alerts;
+
         return $alerts;
     }
+
+    /**
+     * Align alerts to hourly periods for display purposes
+     *
+     * The alerts need to be displayed in the hourly table,
+     * often across multiple hourly columns. We use this
+     * method to process the alert information for display
+     * within those columns.
+     */
+    public function alertsToHourlyPeriods($alerts, $periods)
+    {
+        // Pull out alerts that are relevant to the range
+        // of the current periods
+        $firstPeriodStartTime = \DateTimeImmutable::createFromFormat(
+            \DateTimeInterface::ISO8601_EXPANDED,
+            $periods[0]["timestamp"],
+        );
+
+        $timezone = $firstPeriodStartTime->getTimezone()->getName();
+
+        $lastPeriodEndTime = self::turnToDate(
+            $periods[array_key_last($periods)]["timestamp"],
+            $timezone,
+        );
+        $lastPeriodEndTime = $lastPeriodEndTime->modify("+ 1 hour");
+
+        // Initially weed out any alerts that to not fall
+        // within the overall period range
+        $relevantAlerts = array_filter($alerts, function ($alert) use (
+            &$periods,
+            &$lastPeriodEndTime,
+        ) {
+            $onsetDateTime = self::turnToDate(
+                $alert->onsetRaw,
+                $alert->timezone,
+            );
+            return $onsetDateTime < $lastPeriodEndTime;
+        });
+
+        // We will  respond with a list of alerts from the
+        // incoming alerts that fall within the period range,
+        // annotated with extra information about how
+        // they should be displayed
+        $alertPeriods = [];
+
+        foreach ($relevantAlerts as $currentAlert) {
+            $onsetTime = self::turnToDate(
+                $currentAlert->onsetRaw,
+                $currentAlert->timezone,
+            );
+            $endTime = $this->getEndTimeForAlert($currentAlert);
+            if (!$endTime) {
+                continue; // pass to the next alert, ignoring this one
```

Comment:
@shadkeene @coreypieper @colinmurphy01  Do we need to add a ticket to decide how to address alerts that doesn't have an expiration or end time?

Reply:
yes, using #852 to track